### PR TITLE
Only one access_toke call and show loader in sign_in page when user close and open app after sign in

### PR DESCRIPTION
**Pull request by iamtarun1993**

### CODE REVIEW

File: src/app/auth/sign-in/sign-in.page.ts [hunk 1]
```diff
@@ -252,17 +252,20 @@ export class SignInPage implements OnInit {
     this.emailSet = !!this.fg.value.email;
   }
 
-  async ngOnInit() {
+  ngOnInit() {
     const presentEmail = this.activatedRoute.snapshot.params.email;
     this.fg = this.formBuilder.group({
       email: [presentEmail || '', Validators.compose([Validators.required, Validators.pattern('\\S+@\\S+\\.\\S{2,}')])],
       password: ['', Validators.required]
     });
```
Comment (iamtarun1993) [inline src/app/auth/sign-in/sign-in.page.ts:261]: Converted this promise-based logic into observable logic. And showing one loader when checking the user is already logged in or not.

File: src/app/core/interceptors/httpInterceptor.ts [hunk 2]
```diff
@@ -47,19 +41,6 @@ export class HttpConfigInterceptor implements HttpInterceptor {
     return false;
   }
```
Comment (iamtarun1993) [inline src/app/core/interceptors/httpInterceptor.ts:42]: moved these two methods into `routerAuthService`.

**Review by lakshmisuhas1996 — APPROVED** [review]

**Review by abhishek1234321 — APPROVED** [review]

File: src/app/core/interceptors/httpInterceptor.ts [hunk 3]
```diff
@@ -10,19 +10,21 @@ import {
...
 import * as moment from 'moment';
 import { TokenService } from '../services/token.service';
-import { AuthService } from '../services/auth.service';
 import { RouterAuthService } from '../services/router-auth.service';
 import { DeviceService } from '../services/device.service';
 
 @Injectable()
 export class HttpConfigInterceptor implements HttpInterceptor {
+  private accessTokenCallInProgress = false;
+  private accessTokenSubject: Subject<any> = new BehaviorSubject<any>(null);
```
Comment (Julias0) [inline src/app/core/interceptors/httpInterceptor.ts:26]: private accessTokenSubject = new BehaviorSubject<any>(null);

Also replace any with proper type

File: src/app/core/interceptors/httpInterceptor.ts [hunk 4]
```diff
@@ -72,21 +73,44 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
-            concatMap(authResponse => {
-              return from(this.routerAuthService.newAccessToken(authResponse.access_token))
-            }),
-            concatMap(() => {
-              return from(this.tokenService.getAccessToken())
-            })
-          );
+          if (!this.accessTokenCallInProgress) {
+            this.accessTokenCallInProgress = true;
+            this.accessTokenSubject.next(null);
```
Comment (Julias0) [inline src/app/core/interceptors/httpInterceptor.ts:88]: why are we emitting null here?

File: src/app/core/interceptors/httpInterceptor.ts [hunk 5]
```diff
@@ -72,21 +73,44 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
+            return from(this.tokenService.getRefreshToken()).pipe(
+              concatMap(refreshToken => {
```
Comment (Julias0) [inline src/app/core/interceptors/httpInterceptor.ts:90]: use switchMap instead of concat map

File: src/app/core/interceptors/httpInterceptor.ts [hunk 6]
```diff
@@ -72,21 +73,44 @@ export class HttpConfigInterceptor implements HttpInterceptor {
...
   }
 
-  getAccessToken() {
+  /**
+   * This method get current accessToken from Storage, check if this token is expiring or not.
+   * If the token is expiring it will get another accessToken from API and return the new accessToken
+   * If multiple API call initiated then `this.accessTokenCallInProgress` will block multiple access_token call
+   * Reference: https://stackoverflow.com/a/57638101
+   */
+   getAccessToken(): Observable<string> {
```
Comment (Julias0) [inline src/app/core/interceptors/httpInterceptor.ts:82]: if we are using behaviorSubject

this should be 
```
getAccessToken(): Observable<string>{
  return  this.refreshAccessTokenIfExpired.pipe(
    switchMap(()=> this. accessTokenSubject.asObservable())
   );
}
```

the flag `accessTokenCallInProgress` shouldnt be required - accessTokenSubject should only emit when required

Comment (iamtarun1993) [reply src/app/core/interceptors/httpInterceptor.ts:82]: check this 
https://stackoverflow.com/a/57638101

File: src/app/core/interceptors/httpInterceptor.ts (hunk 3 above)
Comment (Julias0) [reply src/app/core/interceptors/httpInterceptor.ts:26]: dei this is not taken care of

File: src/app/core/interceptors/httpInterceptor.ts (hunk 4 above)
Comment (iamtarun1993) [reply src/app/core/interceptors/httpInterceptor.ts:88]: copied from https://stackoverflow.com/a/57638101

**Review by Julias0 — APPROVED** [review]

**Pull request merged**